src/libs/ajax/User.ts

```diff
@@ -0,0 +1,359 @@
+  userInfo: {
+    userSubjectId: string;
+    userEmail: string;
+  };
+  enabled: {
+    ldap: boolean;
+    allUsersGroup: boolean;
+    google: boolean;
+  };
+}
+
+export interface SamUserTosComplianceStatusResponse {
+  userId: string;
+  userHasAcceptedLatestTos: boolean;
+  permitsSystemUsage: boolean;
+}
+
+export type TerraUserPreferences = {
+  starredWorkspaces?: string;
+} & {};
+
+export interface OrchestrationUserProfileResponse {
+  userId: string;
+  keyValuePairs: TerraUserProfile;
+}
+
+export interface OrchestrationUserPreferLegacyFireCloudResponse {
+  preferTerra: boolean;
+  preferTerraLastUpdated: number;
+}
+
+export interface NihDatasetPermission {
+  name: string;
+  authorized: boolean;
+}
+
+export interface OrchestrationNihStatusResponse {
+  linkedNihUsername: string;
+  datasetPermissions: NihDatasetPermission[];
+  linkExpireTime: number;
+}
+
+export interface BondFenceUrlResponse {
+  url: string;
+}
+
+export interface BondFenceStatusResponse {
+  issued_at: Date;
+  username: string;
+}
+
+export interface EcmLinkAccountResponse {
+  externalUserId: string;
+  expirationTimestamp: Date;
+  authenticated: boolean;
+}
+
+export interface SamInviteUserResponse {
+  userSubjectId: string;
+  userEmail: string;
+}
+
+export interface RexFirstTimestampResponse {
+  timestamp: Date;
+}
+
+// TODO: Remove this as a part of https://broadworkbench.atlassian.net/browse/ID-460
+const getFirstTimeStamp = Utils.memoizeAsync(
+  async (token): Promise<RexFirstTimestampResponse> => {
+    const res = await fetchRex('firstTimestamps/record', _.mergeAll([authOpts(token), { method: 'POST' }]));
+    return res.json();
+  },
+  { keyFn: (...args) => JSON.stringify(args) }
+) as (token: string) => Promise<RexFirstTimestampResponse>;
+
+export const User = (signal?: AbortSignal) => {
+  return {
+    getStatus: async (): Promise<SamUserRegistrationStatusResponse> => {
+      const res = await fetchSam('register/user/v2/self/info', _.mergeAll([authOpts(), { signal }]));
+      return res.json();
+    },
+
+    profile: {
+      get: async (): Promise<OrchestrationUserProfileResponse> => {
+        const res = await fetchOrchestration('register/profile', _.merge(authOpts(), { signal }));
+        return res.json();
+      },
+
+      // We are not calling Thurloe directly because free credits logic was in orchestration
+      set: async (profile: TerraUserProfile): Promise<void> => {
+        interface SetTerraUserProfileRequest {
+          firstName: string;
+          lastName: string;
+          title: string;
+          contactEmail?: string;
+          institute: string;
+          programLocationCity: string;
+          programLocationState: string;
+          programLocationCountry: string;
+          termsOfService?: string;
+          researchArea?: string;
+          department?: string;
+          interestInTerra?: string;
+        }
+        const blankProfileRequest: SetTerraUserProfileRequest = {
```

Comment:
Why do we need a separate `SetTerraUserProfileRequest` type? Will `TerraUserProfile` not work here?

Reply:
I wanted to make a distinction between the request type and the type we use in the UI. There are alot of fields that are in the TerraUserProfile which are not a part of the `register/profile` Orch request. This is also confusing because 
`https://firecloud-orchestration.dsde-dev.broadinstitute.org/#/Profile/setProfile` and `https://firecloud-orchestration.dsde-dev.broadinstitute.org/#/Profile/getAll` do not operate on the same resource even though it seems like they would. `setProfile` takes a specific "Profile" object with specific fields (behind the scenes this puts those "Profile" fields as key values in Thurloe). `getAll` gives every keyValue pair associated with the user. Some of those things are the same fields from the "Profile" object but there can be more. 

There is a general smell going on that profile and preferences are being used in the UI sort of interchangeably (because they are just key value pairs in Thurloe). I would like to get it to a place where we can abstract away the arbitrary key value pairs and only use the fields that the UI needs. We should not need to replicate Thurloe's bad implementation with an equally bad implementation in the UI.

Reply:
Ok, but that sounds like `set` should accept a `SetTerraUserProfileRequest`. Currently, it accepts a `TerraUserProfile` and `SetTerraUserProfileRequest` is only used for the defaults in `blankProfileRequest`.